### PR TITLE
166267668 setup node and express

### PR DESCRIPTION
**Pull request by allebd**
#### Description
Setup node and express

#### Type of change
- install node
- install express

#### How Has This Been Tested?
NA

#### Checklist:
NA

#### Pivotal
#166267668

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import express from 'express';
+
+const app = express(); // calling an instance of express
+const PORT = 5000; // setup PORT to be used
```
Comment (Adekoreday) [inline server/index.js]: great job, PORT might preferably be defined as an environmental variable like
PORT = 5000 || process.env.PORT

**Review by Adekoreday — COMMENTED** [review]
Great Job , added a comment

**Review by Lundii — APPROVED** [review]
Approved

**Review by Adekoreday — APPROVED** [review]

**Comment by henryade** [comment]
Yeah good job.

File: server/index.js (hunk 1 above)
Comment (allebd) [reply server/index.js]: Noted

Comment (KvNGCzA) [reply server/index.js]: @allebd process.env.PORT should come first.

File: server/index.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+app.get('/', (request, response) => {
+    response.status(200).send('Hello Books');
+});
+
+app.listen(PORT, () => {
+    console.log(`Server is running on PORT ${PORT}`);
+});
```
Comment (KvNGCzA) [inline server/index.js]: Add a new line at the end of this file

**Review by KvNGCzA — CHANGES_REQUESTED** [review]
Kindly add error handlers to the server file.

File: server/index.js (hunk 1 above)
Comment (allebd) [reply server/index.js]: Working on it

File: server/index.js (hunk 2 above)
Comment (allebd) [reply server/index.js]: Noted, working on it

File: server/index.js (hunk 1 above)
Comment (allebd) [reply server/index.js]: Done

File: server/index.js (hunk 2 above)
Comment (allebd) [reply server/index.js]: Done

File: server/index.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        response.status(200).send('Hello Books');
+    } catch (e){
+        response.status(404).send('Address not found');  
+    }
+});
+
+app.listen(PORT, () => {
+    try{
+        console.log(`Server is running on PORT ${PORT}`);
+    } catch (e){
```
Comment (Ukhu) [inline server/index.js]: I think you should use 'error' here for better readability

File: server/index.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import express from 'express';
+
+const { PORT } = process.env; // setup PORT to be used
```
Comment (KvNGCzA) [inline server/index.js]: How will you extract the `PORT` from the `environment`?

File: server/index.js (hunk 3 above)
Comment (henryade) [reply server/index.js]: Everything looks good to me

Comment (allebd) [reply server/index.js]: Noted, Working on it

Comment (allebd) [reply server/index.js]: Done

File: server/index.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+const app = express(); // calling an instance of express
+
+app.get('/', (request, response) => {
+    try{
+        response.status(200).send('Hello Books');
+    } catch (error){
```
Comment (KvNGCzA) [inline server/index.js]: What I mean by error handlers are handlers that catch routes that do not exist and return a 404 or 500 depending on the error.

**Review by KvNGCzA — CHANGES_REQUESTED** [review]

File: server/index.js (hunk 4 above)
Comment (allebd) [reply server/index.js]: This has just been corrected

Comment (KvNGCzA) [reply server/index.js]: @allebd What I'm asking is where will be `PORT` be extracted from, you do not have `dotenv` installed, so how will `process.env` be called

File: server/index.js (hunk 5 above)
Comment (allebd) [reply server/index.js]: Done, treated accordingly

File: server/index.js (hunk 4 above)
Comment (allebd) [reply server/index.js]: dotenv has just been added accordingly

**Comment by henryade** [comment]
it looks good. you can add a .env.example file though.

File: server/index.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import { config } from 'dotenv';
+import logger from 'morgan';
+import cors from 'cors';
+
+config();
+
+const { PORT = 5000 } = process.env; // setup port to be used
+const app = express(); // calling an instance of express
+
+const isDevelopment = process.env.NODE_ENV === 'development';
```
[the remark this answers is not in the document]
Comment (Adekoreday) [reply server/index.js]: i dont think you need to do this defining port = process.env.PORT || 5000 will automatically assign environment variable if it exist and resolve to 5000 if it doesn't exist

File: server/index.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+app.use(cors());
+
+// index route
+app.get('/', (request, response) => {
+    response.status(200).send('Hello Books');
+});
+
+// catch 404 and forward to error handler
+app.use((request, response, next) => {
+    const error = new Error('Not Found');
```
[the remark this answers is not in the document]
Comment (Adekoreday) [reply server/index.js]: to handle 404 error a better approach is to use
app.get( * , req, res => appropriate error handling response );
go through this 
https://www.freecodecamp.org/news/really-really-basic-routing-in-nodejs-with-express-d7cad5e3f5d5/

File: server/index.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
+
+import express, { json, urlencoded } from 'express';
+import { config } from 'dotenv';
+import logger from 'morgan';
+import validator from 'express-validator';
+import cors from 'cors';
+
+config();
+
+const { PORT = 5000 } = process.env; // setup port to be used
```
Comment (KvNGCzA) [inline server/index.js]: You should get rid of this `5000` and include it in your .env file

**Review by KvNGCzA — CHANGES_REQUESTED** [review]

**Review by KvNGCzA — CHANGES_REQUESTED** [review]
How do I start this server?

File: server/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (allebd) [reply server/index.js]: Done, treated accordingly

File: server/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (allebd) [reply server/index.js]: Done, Treated accordingly

File: server/index.js (hunk 8 above)
Comment (allebd) [reply server/index.js]: Done, treated accordingly

File: server/index.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+app.use(urlencoded({ extended: false }));
+app.use(validator());
+app.use(cors());
+
+// index route
+app.get('/', (request, response) => {
+    response.status(200).send('Hello Books');
+});
+
+app.use('*', (request, response) => {
```
[the remark this answers is not in the document]
Comment (Adekoreday) [reply server/index.js:24]: ok good

**Review by KvNGCzA — APPROVED** [review]
Get one more approval so your PR can be merged

**Review by Ukhu — APPROVED** [review]
Good. Much better

File: server/index.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
+
+import express, { json, urlencoded } from 'express';
+import { config } from 'dotenv';
+import logger from 'morgan';
+import validator from 'express-validator';
+import cors from 'cors';
+
+config();
+
+const { PORT } = process.env; // setup port to be used
```
[the remark this answers is not in the document]
Comment (Adekoreday) [reply server/index.js]: approved

**Pull request merged**